### PR TITLE
[RFC]Timeaware GPIO API and driver

### PR DESCRIPTION
**Pull request by sknaduva**
Submitting this PR for review for Timeaware GPIO driver. This is a new driver.

This PR includes 
a. Draft API set ( subjected to review)
b. dts node
c. implementation of driver (subjected to modification after API review)

Timeaware GPIO or TGPIO is a new IP found in various Intel platforms(eg. Elkhartlake, Raptorlake, etc.) which allows for time synchronization between systems without any other means of synchronization such as PTP or NTP. 
Using this IP.

a) a pin can be configured to generate synchronized output pulses based on
a system timer.
b) Can detect pulses on a pin and timestamp the arrival of the pulse.

Different clocks on different system can "skew" (faster or slower based on the doping variation of clocking crystal) over time. Timeaware GPIO allows sync time between two systems using hardware pins by generating pulses based on A's clock on System A and timestamping arrival of same pulse based on B's clock system B.

eg. A is master/primary clock, and B is secondary clock.

System A======XX=========>A's TGPIO pin ------------------ B's TGPIO pin=======YY======> System B.

XX generate pulse every second.
YY timestamp when arrived.

Calculate the skew and adjust the B's system clock.

Please refer 22.18 of https://cdrdv2.intel.com/v1/dl/getContent/636112 for
details about the IP.

### CODE REVIEW

File: drivers/timeaware_gpio/timeaware_gpio_intel.c [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	if (event_polarity == 0) {
+		val = CTL_EP_RISING_EDGE;
+	} else if (event_polarity == 1) {
+		val = CTL_EP_FALLING_EDGE;
+	} else {
+		val = CTL_EP_TOGGLE_EDGE;
+	}
+
+	/* Configure direction = input */
+	val = val | CTL_DIR;
```
Comment (softwarecki) [inline drivers/timeaware_gpio/timeaware_gpio_intel.c:165]: val |= CTL_DIR;

File: drivers/timeaware_gpio/timeaware_gpio_intel.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+	return 0;
+}
+
+static int tagpio_intel_config_external_timestamp(const struct device *dev,
+							 uint32_t pin,
+							 uint32_t event_polarity)
+{
+	mm_reg_t addr = regs(dev);
+	uint32_t val = 0;
```
Comment (softwarecki) [inline drivers/timeaware_gpio/timeaware_gpio_intel.c:146]: Initialization of the `val` variable is superfluous.

File: drivers/timeaware_gpio/timeaware_gpio_intel.c [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	val = repeat_interval & UINT32_MASK;
+	sys_write32(val, addr + PIV31_0);
+
+	/* Configure COMPV */
+	val = (start_time >> UINT32_SIZE) & UINT32_MASK;
+	sys_write32(val, addr + COMPV63_32);
+	val = start_time & UINT32_MASK;
+	sys_write32(val, addr + COMPV31_0);
+
+	val = 0;
```
Comment (softwarecki) [inline drivers/timeaware_gpio/timeaware_gpio_intel.c:127]: Just init `val` with CTL_EN

File: drivers/timeaware_gpio/timeaware_gpio_intel.c [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+static mm_reg_t regs(const struct device *dev)
+{
+	return DEVICE_MMIO_NAMED_GET(dev, reg_base);
+}
+
+static int tagpio_intel_get_time(const struct device *dev,
+					uint64_t *time)
+{
+	*time = sys_read32(regs(dev) + ART_L);
+	*time = *time + ((uint64_t)sys_read32(regs(dev) + ART_H) << UINT32_SIZE);
```
Comment (softwarecki) [inline drivers/timeaware_gpio/timeaware_gpio_intel.c:71]: *time += ....;
val |= ... in all other places

**Review by softwarecki — CHANGES_REQUESTED** [review]
Why did you decide to put the driver in a new directory? Wouldn't the gpio directory be appropriate?

**Comment by marc-hb** [comment]
Off-topic sorry but I think something just started going wrong with the `zephyr-bot`. This is the second non-audio PR where zephyr-bot adds the [platform: intel ADSP](https://github.com/zephyrproject-rtos/zephyr/labels/platform%3A%20Intel%20ADSP) label and audio reviewers for no obvious reason (the other one was #55841).

Where is the configuration/code of the `zephyr-bot`?

**Comment by marc-hb** [comment]
Answering myself: being fixed in #55848. Thanks @MaureenHelm

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please add further details on why a new API is needed, what problems does it solve, whether this is a common class of devices (more than one vendor currently makes hardware, that fits this API), etc.

**Comment by gmarull** [comment]
It looks like there's a Linux PTP driver for same/similar IP authored by Intel: https://github.com/intel/linux-intel-quilt/blob/mainline-tracking-v5.11-yocto-210223T083754Z/patches/0005-Add-support-for-PMC-Time-Aware-GPIO-TGPIO-hardware-t.tgpio

any reason this is being done as a new API here? why not use ptp?

**Comment by sknaduva** [comment]
> It looks like there's a Linux PTP driver for same/similar IP authored by Intel: https://github.com/intel/linux-intel-quilt/blob/mainline-tracking-v5.11-yocto-210223T083754Z/patches/0005-Add-support-for-PMC-Time-Aware-GPIO-TGPIO-hardware-t.tgpio
> 
> any reason this is being done as a new API here? why not use ptp?

Above link has a ptp driver implementation, but in intel mainline. It didn't succeed to make it to Linus' tree. But certainly we can give it a try.

**Comment by carlescufi** [comment]
Architecture WG:

- @sknaduva states that this 
- @henrikbrixandersen mentions that we discussed whether we could put it in the PTP API in Zephyr. @sknaduva says the option has been explored but rejected because the API doesn't fit
- @nashif questions whether Timeaware GPIO is really an implementation of a PTP clock
- @henrikbrixandersen asks whether we should extend or modify the ptp_clock API to cover the functionality shown here
- This doesn't seem like a good fit for the ptp_clock API
- @galak suggests creating a new `swclock` driver class, alternative is to place it under `misc`

**Review by cfriedt — CHANGES_REQUESTED** [review]
Personally, I think Time aware GPIO is great, conceptually. There are many fields where picoseconds truly matter, and this gives users the ability to meet very tight timing constraints.

However, I think that the desired results can be achieved by simply extending the GPIO API. By that, I mean
* extend the gpio controller DT bindings (probably add a "timeaware" bool)
* extend the gpio driver API (wrap the regular gpio api in a tag gpio api and add custom methods / function pointers)
* possibly pass in additional driver-specific flags to existing gpio api calls

**Comment by jhedberg** [comment]
Superseded by #59825

**Pull request closed without merging**